### PR TITLE
modify host name if needed

### PR DESCRIPTION
**Pull request by zqingqing1**
Simply comparing req.url.host with transport host is not enough, we need to consider the host with portNumber cases. In this PR, I considered 4 cases and modify the transport host if needed.

### CODE REVIEW

File: utils.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package armbalancer
+
+import (
+	"strings"
+)
+
+func getTransportHostToCompare(reqHost, transportHost string) string {
```
Comment (jveski) [inline utils.go:7]: I'd prefer to keep this in armbalancer.go and make it a member of `recyclableTransport` so we can reference `t.host` directly without needing to pass it as an arg.

File: armbalancer.go [hunk 2]
```diff
@@ -113,8 +113,9 @@ func newRecyclableTransport(id int, parent *http.Transport, host string, recycle
 }
 
 func (t *recyclableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
-	if req.URL.Host != t.host {
-		return nil, fmt.Errorf("host %q is not supported by the configured ARM balancer", req.URL.Host)
+	host := getTransportHostToCompare(req.URL.Host, t.host)
+	if req.URL.Host != host {
+		return nil, fmt.Errorf("host %q is not supported by the configured ARM balancer, supported host name is %q", req.URL.Host, host)
```
Comment (jveski) [inline armbalancer.go:118]: I think `req.URL.Host, host)` should be `req.URL.Host, t.host)`

**Review by jveski — CHANGES_REQUESTED** [review]
Great find — thanks for writing the fix. Just a few comments on style/organization.

File: utils_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+type testCase struct {
+	name      string
+	reqHost   string
+	transHost string
+	expected  string
+}
+
+func TestGetTransportHostToCompare(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (zqingqing1) [reply utils_test.go:14]: thanks

File: utils.go (hunk 1 above)
Comment (zqingqing1) [reply utils.go:7]: I found it will have race condition to read and update t.host by multiple go routines during tests.

Comment (zqingqing1) [reply utils.go:7]: no worries, found a way that no need to update t.host

File: armbalancer.go [hunk 4]
```diff
@@ -113,8 +113,9 @@ func newRecyclableTransport(id int, parent *http.Transport, host string, recycle
 }
 
 func (t *recyclableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
-	if req.URL.Host != t.host {
-		return nil, fmt.Errorf("host %q is not supported by the configured ARM balancer", req.URL.Host)
+	host := getTransportHostToCompare(req.URL.Host, t.host)
```
[the remark this answers is not in the document]
Comment (zqingqing1) [reply armbalancer.go]: synced offline, and fixed it

**Review by jveski — APPROVED** [review]

**Pull request merged**